### ISSUE TITLE
Provide a standard way to recuperate mission logs from moles.

### ISSUE THREAD

**Issue by kaifox** [molr/molr#20]
At a first glance, I would tend to try something like the following in the mole API:
```
Flux<Log> logFor(MissionHandle handle); 
```
A Log would contain a list of log lines, where each line would have information of a log statement, enriched by molr specific information.....e.g. something like:

```
[
  {
    "time" : "2018-12-01 18:45:12.000",
    "missionHandle" : "mid0",
    "strandId" : "strand01",
    "blockId" : "block0",
    "message" : "This is a log message"
  },
  { ...}
  ...
] 
```

Ideally, similar to the MissionOutput, the Log would grow during mission execution, and always deliver the latest snapshot. On subscription (even after finishing the message), it would deliver one update with the latest snapshot.

Every mole implementation would have to take care to collect the relevant information through its standard logging environment. In the java case this might most likely be slf4j (or an implementation of it).... Something similar in python probably...

The tricky part here is of course on the mole side, how to collect this information .... No clear idea at this side 

When investigating these concept for acctesting, Andrej mentioned a concept from log4j: MDC, which could be helpful to achieve this.

**Comment by swecooo** [comment]
Yes, we were able to collect the logs for individual missions ran on java moles by using the MDC. The implementation adds an appender to the root logger and the appender filters only for logs with a specific MDC. Since adding an appender is not possible via slf4j only, we had to use a logging library on the classpath (we used log4j).

We went for this solution mainly because it enables to collect also the logs of 3rd party libraries, as these are (usually) logging using slf4j.

```java
String uuid = UUID.randomUUID().toString();
StringArrayAppender appender = new StringArrayAppender();
appender.addFilter(new MdcFilter(uuid));
org.apache.log4j.Logger.getRootLogger().addAppender(appender);
MDC.put(UUID_MDC_KEY, uuid);
try {
    // run mission code
} catch (RuntimeException e) {
    LOGGER.error("Exception thrown while executing mission", e);
    throw e;
} finally {
    MDC.remove(UUID_MDC_KEY);
    org.apache.log4j.Logger.getRootLogger().removeAppender(appender);
    out.emit(LOG_PLACEHOLDER, appender.getLogs());
}
```

The `MdcFilter` then looks like:
```java
private static class MdcFilter extends Filter {

    private final String uuid;

    public MdcFilter(String uuid) {
        this.uuid = uuid;
    }

    @Override
    public int decide(LoggingEvent event) {
        if (uuid.equals(event.getMDC(UUID_MDC_KEY))) {
            return ACCEPT;
        } else {
            return DENY;
        }
    }
}
```

**Comment by swecooo** [comment]
For python, this could work: https://github.com/AFriemann/mdc

**Comment by andreacalia** [comment]
@HimanshuSahu31 thank you very much for your prototype in the two pull requests #37 and #38 !
After a quick chat with @michi42 , we would like to propose a rethink on how we can handle the logs. As shown in the two pull requests, it is technically possible to have the logs nicely republished as stream so the users can consume them.
Unfortunately, using custom objects to intercepts them does not feel very natural since in the actual mission code one could forget to do so. Also any 3rd party log will not be captured, as pointed out by @swecooo in his comment :)
Using MDC is an option, but the question is how to do it properly in the framework so that is works for the general case.
What I can see at the moment, is that every strand is executed on an instance of `ConcurrentStrandExecutor`. Inside, it uses an `Executors.newSingleThreadExecutor(...)` to spawn a new thread. One proposal could be to add the correct MDC information at this point so that every mission block this strand executes will have the correct information. Also, this will work everywhere at the moment for every normal call to a slf4j logger.
Then, we have to actually decide which implementation we use (logback or log4j) to provide a custom appender that filters the logs depending on the MDC and store/republish them as stream. This part does not differ too much from the implementation of @HimanshuSahu31 or @swecooo ..

The issue with this approach is that if a mission instantiates its own threads then it is possible that the MDC information will be lost depending if the thread local variables are propagated or not. I personally think this is acceptable if documented since it is anyway not recommended to spawn custom threads in a mission...

What do you think about this? I would like to thank you guys for trying this out, I think we're converging toward a nice solution :)
